### DISCUSSION TITLE
Is there a way to get fingerprints as a Python bytes object directly without BitVectToFPSText and fromhex?

### DISCUSSION THREAD

**Discussion by drassokhin** [rdkit/rdkit#5741] [Q&A] [answered]
Is there any way to get fingerprints as a Python bytes object directly without this BitVectToFPSText to string and fromhex to bytes trip?
```Python
    fp = AllChem.GetMorganFingerprintAsBitVect(mol, 2, nBits=VECTOR_DIMENSION)
    hex_fp = DataStructs.BitVectToFPSText(fp)
    vec = bytes.fromhex(hex_fp)
```

**Comment by ptosco** [comment, marked as answer]
@drassokhin `DataStructs.BitVectToBinaryText(fp)`.

**Comment by ptosco** [comment]
```
mol = Chem.MolFromSmiles("O=C(O)C(c1ccc(cc1)CCN4CCC(c2nc3ccccc3n2CCOCC)CC4)(C)C")
fp = AllChem.GetMorganFingerprintAsBitVect(mol, 2, 1024)
DataStructs.BitVectToBinaryText(fp) == bytes.fromhex(DataStructs.BitVectToFPSText(fp))
True
```

**Reply by drassokhin** [reply]
@ptosco Thanks, it works now! The interface is a bit confusing though. I think it would make sense to define BitVectToBinaryText as a member of the ExplicitBitVect type (and call it ToBitset or ToBitarray, or ToBytes, for example, to distinguish it from the ToBinary method).
BTW, I looked at the implementation of  BitVectToBinaryText in rdkit/Code/DataStructs/BitOps.cpp just out of curiosity, yes, it does exactly what I would expect it to do. I have a minor suggestion though:

```C++

template <typename T1>
std::string BitVectToBinaryText(const T1& bv1) {
  // Original version was: std::string res(bv1.getNumBits() / 8 + (bv1.getNumBits() % 8 ? 1 : 0), 0);
  // Can simplify as follows: -- D.R.
 std::string res((bv1.getNumBits() + 7) / 8, 0)
  ................
}
```

**Reply by ptosco** [reply]
Thanks @drassokhin for accepting the answer and for cleaning up the page!
